sys/include/net/asymcute.h

```diff
@@ -0,0 +1,568 @@
+    ASYMCUTE_SUBSCRIBED,            /**< client was subscribed to topic */
+    ASYMCUTE_UNSUBSCRIBED,          /**< client was unsubscribed from topic */
+};
+
+/**
+ * @brief   Forward type declaration for connections contexts
+ */
+typedef struct asymcute_con asymcute_con_t;
+
+/**
+ * @brief   Forward type declaration for request contexts
+ */
+typedef struct asymcute_req asymcute_req_t;
+
+/**
+ * @brief   Forward type declaration for subscription contexts
+ */
+typedef struct asymcute_sub asymcute_sub_t;
+
+/**
+ * @brief   Forward type declaration for topic definitions
+ */
+typedef struct asymcute_topic asymcute_topic_t;
+
+/**
+ * @brief   Forward type declaration for last will definitions
+ */
+typedef struct asymcute_will asymcute_will_t;
+
+/**
+ * @brief   Event callback used for communicating connection and request related
+ *          events to the user
+ *
+ * @param[in] req       pointer to the request context that triggered the event,
+ *                      may be NULL of unsolicited events
+ * @param[in] evt_type  type of the event
+ */
+typedef void(*asymcute_evt_cb_t)(asymcute_req_t *req, unsigned evt_type);
+
+/**
+ * @brief   Callback triggered on events for active subscriptions
+ *
+ * @param[in] sub       pointer to subscription context triggering this event
+ * @param[in] evt_type  type of the event
+ * @param[in] data      incoming data for PUBLISHED events, may be NULL
+ * @param[in] len       length of @p data in bytes
+ * @param[in] arg       user supplied argument
+ */
+typedef void(*asymcute_sub_cb_t)(const asymcute_sub_t *sub, unsigned evt_type,
+                                 const void *data, size_t len, void *arg);
+
+/**
+ * @brief   Context specific timeout callback, only used internally
+ *
+ * @internal
+ *
+ * @param[in] con       connection context for this timeout
+ * @param[in] req       request that timed out
+ *
+ * @return  Event type to communicate to the user
+ */
+typedef unsigned(*asymcute_to_cb_t)(asymcute_con_t *con, asymcute_req_t *req);
+
+/**
+ * @brief   Asymcute request context
+ */
+struct asymcute_req {
+    struct asymcute_req *next;  /**< the requests list entry */
+    mutex_t lock;               /**< synchronization lock */
+    asymcute_con_t *con;        /**< connection the request is using */
+
+    event_callback_t to_evt;    /**< timeout event */
+    event_timeout_t to_timer;   /**< timeout timer */
+    uint8_t retry_cnt;          /**< retransmission counter */
+
+    uint16_t msg_id;            /**< used message id for this request */
+    size_t data_len;            /**< length of the request packet in byte */
+    uint8_t data[ASYMCUTE_BUFSIZE]; /**< buffer holding the request's data */
+
+    asymcute_to_cb_t cb;        /**< internally used callback */
+    void *arg;                  /**< internally used additional state */
+};
+
+/**
+ * @brief   Asymcute connection context
+ */
+struct asymcute_con {
+    uint8_t state;              /**< connection state */
```

Comment:
IMO this would be premature optimization, as the size of this struct is quite significant anyway (and contains 'wasted' fill bytes depending on the configuration)... In this case I actually think that there is a better gain by logically grouping the fields in this struct.

Reply:
At what point becomes sensible optimization premature. I actually don't see the logical connection of `state` to `lock` or `last_id` (message ID) to `client_id` (except for the fact that they are both called ID). The only field I agree with is `keepalive_retry_cnt`, but if you move `state` below and `last_id` below that, you still have a logical grouping *and* waste-less packing ;-).

Reply:
ok ok